src/node.cc

```diff
@@ -4331,142 +4349,307 @@ Local<Context> NewContext(Isolate* isolate,
+ */
+class CmdArgs {
+ public:
+  /**
+   * @brief CmdArgs creates valid argc and argv variables from a program name
+   * and arguments.
+   *
+   * The argv buffer is a contiguous, adjacent char buffer and contains the
+   * program name as its first item followed by the provided arguments. argc
+   * is the number of arguments + 1 (the program name).
+   * The resulting argv buffer should not be modified.
+   *
+   * @param program_name the name of the executable
+   * @param arguments the arguments for the program
+   */
+  CmdArgs(const std::string& program_name,
+          const std::vector<std::string>& arguments)
+    : argc(0),
+      argv(nullptr) {
+    size_t total_size = 0;
+    total_size += program_name.size() + 1;
+    for (const auto& argument : arguments) {
+      total_size += argument.size() + 1;
+    }
 
-  {
-    SealHandleScope seal(isolate);
-    bool more;
-    PERFORMANCE_MARK(&env, LOOP_START);
-    do {
-      uv_run(env.event_loop(), UV_RUN_DEFAULT);
+    std::vector<std::size_t> offsets;
+    argument_data_.reserve(total_size);
+    offsets.push_back(argument_data_.size());
+    argument_data_ += program_name;
+    argument_data_ += static_cast<char>(0x0);
+    for (const auto& argument : arguments) {
+      offsets.push_back(argument_data_.size());
+      argument_data_ += argument;
+      argument_data_ += static_cast<char>(0x0);
+    }
 
-      v8_platform.DrainVMTasks(isolate);
+    argument_pointers_.resize(offsets.size());
+    for (std::size_t i=0; i < argument_pointers_.size(); ++i) {
+      argument_pointers_[i] = argument_data_.data() + offsets[i];
+    }
+    argc = argument_pointers_.size();
+    argv = argument_pointers_.data();
+  }
+
+  ~CmdArgs() = default;
+
+  /**
+   * @brief argc is the number of arguments + 1 (the program name)
+   */
+  int argc;
+  /**
+   * @brief argv is an array containing pointers to the arguments (and the
+   * program name), it should not be modified
+   */
+  const char** argv;
+
+ private:
+  /**
+   * @brief argument_data contains the program name and the arguments separated
+   * by null bytes
+   */
+  std::string argument_data_;
+  /**
+   * @brief argument_pointers contains pointers to the beginnings of the
+   * strings in argument_data
+   */
+  std::vector<const char*> argument_pointers_;
+};
 
-      more = uv_loop_alive(env.event_loop());
-      if (more)
-        continue;
+class HandleScopeHeapWrapper {
+ public:
+  explicit HandleScopeHeapWrapper(v8::Isolate* isolate)
+    : scope_(isolate) { }
+  private:
+  HandleScope scope_;
+};
+
+ArrayBufferAllocator* allocator;
+Isolate::CreateParams params;
+Locker* locker;
+IsolateData* isolate_data;
+Isolate::Scope* isolate_scope;
+Local<Context> context;
+Context::Scope* context_scope;
+bool request_stop = false;
+CmdArgs* cmd_args = nullptr;
+bool _event_loop_running = false;
+v8::Isolate* _isolate = nullptr;
+Environment* _environment = nullptr;
+HandleScopeHeapWrapper* _handle_scope_wrapper = nullptr;
```

Comment:
I think almost none of these should be global state; if you need to access any of these as an embedder, we should provide a way to get them from the current environment.

Reply:
We need to preserve those variables between calls to `node::Initialize()`, `node::Run()`, `node::Evaluate()`, ... more calls ..., and lastly, `node::Deinitialize()`.
If we don't save references to these variables, we can't `delete` them in `node::Deinitialize()`. However, these variables aren't public and not accessible by the embedder. We only specifically expose the isolate and environment via getters: https://github.com/nodejs/node/pull/19005/files#diff-971605a9f5d10dae32c48c59f595384eR25

However, we see potential to further reduce the amount of references needed by looking into the variables and inter-object relationships already present. For example, we can certainly remove our `_isolate` variable and instead use the already existing `node_isolate`. We've opened an issue in our repo to further investigate these: https://github.com/hpicgs/node/issues/81

Reply:
@addaleax We opted to maintain the functionality of the existing `Start(int, char**)` implementation, where *one* isolate, *one* environment and event_loop instance were created during node’s initialization. Therefore, we do create those instances in our Initialize method and safe them as globals. The embedder still has the possibility to create further isolates/environments with Node.js' already existing methods. 

However, we see the need to not treat it as a singleton. Therefore, we extended our interface by adding new overloaded methods which do accept an environment parameter. For simple use cases, the user does not have to provide an environment and the default environment created by `Initialize`/`Start` is used.

When we started working on extending the library interface, we assumed the main work to be executed in C++ since Node.js is used as a library, therefore we did not think about using more than one environment.
Are multiple environments a common use case?